File: ekf_ws/src/pgs_pkg/src/pgs_node.py
#!/usr/bin/env python3

"""
Use commands and measurements to perform Pose-Graph SLAM.
This estimates the map as well as all current and pase veh poses.
"""
import rospy
from data_pkg.msg import Command
from geometry_msgs.msg import Vector3
from std_msgs.msg import Float32MultiArray
from data_pkg.msg import PGSState

# import params script.
import rospkg
import importlib.util, sys
spec = importlib.util.spec_from_file_location("module.name", rospkg.RosPack().get_path('data_pkg')+"/src/import_params.py")
module = importlib.util.module_from_spec(spec)
sys.modules["module.name"] = module
spec.loader.exec_module(module)
# set params.
params = module.Config.params

############ GLOBAL VARIABLES ###################
state_pub = None
cmd_queue = []; lm_meas_queue = []
pgs = None
#################################################

def pgs_iteration(event):
    """
    Run one timestep of pose-graph slam using the next cmd and measurement.
    """
    # skip if params not read yet or there's no prediction or measurement.
    if pgs is None or len(cmd_queue) < 1 or len(lm_meas_queue) < 1:
        return
    # pop the next data off the queue.
    cmd = cmd_queue.pop(0)
    lm_meas = lm_meas_queue.pop(0)

def get_init_veh_pose(msg):
    global cur
    # get the vehicle's starting pose.
    cur = [msg.x, msg.y, msg.z]

def get_cmd(msg):
    # get commanded velocities.
    global cmd_queue
    cmd_queue.append([msg.fwd, msg.ang])

def get_landmarks(msg):
    # get measurement of landmarks.
    # format: [id1,range1,bearing1,...idN,rN,bN]
    global lm_meas_queue
    lm_meas_queue.append(msg.data)

def main():
    global state_pub
    rospy.init_node('pgs_node')

    # subscribe to the initial veh pose.
    rospy.Subscriber("/truth/init_veh_pose",Vector3, get_init_veh_pose, queue_size=1)
    # subscribe to commands sent to the vehicle.
    rospy.Subscriber("/command", Command, get_cmd, queue_size=1)
    # subscribe to landmark detections.
    rospy.Subscriber("/landmark", Float32MultiArray, get_landmarks, queue_size=1)

    # create publisher for the current state.
    state_pub = rospy.Publisher("/state/ekf", PGSState, queue_size=1)

    rospy.Timer(rospy.Duration(params["DT"]), pgs_iteration)
    rospy.spin()


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: ekf_ws/src/pgs_pkg/src/pgs_node.cpp

#include <ros/ros.h>
#include "geometry_msgs/Vector3.h"
#include "std_msgs/Float32MultiArray.h"
#include "data_pkg/PGSState.h"
#include "data_pkg/Command.h"
#include <eigen3/Eigen/Dense>
#include <vector>
#include <cmath>
#include <queue>
#include <iostream>
#include <string>
#include <fstream>
#include <ros/package.h>

#include <g2o/core>


// define queues for messages.
std::queue<data_pkg::Command::ConstPtr> cmdQueue;
std::queue<std_msgs::Float32MultiArray::ConstPtr> lmMeasQueue;
// define EKF state publisher.
ros::Publisher statePub;

float readParams() {
    // read config parameters from file.
    float DT;
    std::string delimeter = " = ";
    std::string line;
    // std::string pkgPath = ros::package::getPath("data_pkg");
    std::string pkgPath = "/home/kevin-robb/Projects/live_ekf_slam/ekf_ws/src/data_pkg";
    std::ifstream paramsFile(pkgPath+"/config/params.txt", std::ifstream::in);
    std::string token;
    // read all lines from file.
    while (getline(paramsFile, line)) {
        // only save things we need.
        token = line.substr(0, line.find(delimeter));
        if (token == "DT") {
            line.erase(0, line.find(delimeter)+delimeter.length());
            DT = std::stof(line);
        }
    }
    // close file.
    paramsFile.close();
    return DT;
}

void initCallback(const geometry_msgs::Vector3::ConstPtr& msg) {
    // receive the vehicle's initial position.
    float x_0 = msg->x;
    float y_0 = msg->y;
    float yaw_0 = msg->z;
    // init the EKF.
    // ekf.init(x_0, y_0, yaw_0);
}

void ekfIterate(const ros::TimerEvent& event) {
    // perform an iteration of the EKF for this timestep.
    if (cmdQueue.empty() || lmMeasQueue.empty()) {
        return;
    }
    // get the next timestep's messages from the queues.
    data_pkg::Command::ConstPtr cmdMsg = cmdQueue.front();
    cmdQueue.pop();
    std_msgs::Float32MultiArray::ConstPtr lmMeasMsg = lmMeasQueue.front();
    lmMeasQueue.pop();
    // call the EKF's update function.
    // ekf.update(cmdMsg, lmMeasMsg);
    // get the current state estimate.
    // data_pkg::EKFState stateMsg = ekf.getState();
    // publish it.
    // statePub.publish(stateMsg);
}

void cmdCallback(const data_pkg::Command::ConstPtr& msg) {
    // receive an odom command and add to the queue.
    cmdQueue.push(msg);
}

void lmMeasCallback(const std_msgs::Float32MultiArray::ConstPtr& msg) {
    // receive a landmark detection and add to the queue.
    lmMeasQueue.push(msg);
}

int main(int argc, char **argv) {
    ros::init(argc, argv, "ekf_node");
    ros::NodeHandle node("~");

    // read config parameters.
    float DT = readParams();
    // get the initial veh pose and init the ekf.
    ros::Subscriber initSub = node.subscribe("/truth/init_veh_pose", 1, initCallback);

    // subscribe to EKF inputs.
    ros::Subscriber cmdSub = node.subscribe("/command", 100, cmdCallback);
    ros::Subscriber lmMeasSub = node.subscribe("/landmark", 100, lmMeasCallback);
    // publish EKF state.
    // statePub = node.advertise<data_pkg::EKFState>("/state/ekf", 1);

    // timer to update EKF at set frequency.
    ros::Timer ekfIterationTimer = node.createTimer(ros::Duration(DT), &ekfIterate, false);

    ros::spin();
    return 0;
}